kevm-pyk/src/kevm_pyk/utils.py

```diff
@@ -130,15 +130,57 @@ def kevm_apr_prove(
         )
     except Exception as e:
         _LOGGER.error(f'Proof crashed: {_cfgid}\n{e}', exc_info=True)
-        return False
+        return False, None
 
     failure_nodes = _cfg.frontier + _cfg.stuck
     if len(failure_nodes) == 0:
         _LOGGER.info(f'Proof passed: {_cfgid}')
-        return True
+        return True, None
     else:
         _LOGGER.error(f'Proof failed: {_cfgid}')
-        return False
+        failure_log: list[str] | None = None
+        failure_log = print_failure_info(proof.kcfg, proof.id, kcfg_explore)
+        return False, failure_log
```

Comment:
I would still move the printing of the failure info out of this method, so that it doesn't need to return a tuple. At the call site, you can just check the result and decide whether to call `print_failure_info`. The call site has access to the `proof` (because it's passed in by reference), and to the `kcfg_explore`, so it has everything needed to call `print_failure_info` function.